libs/core/include/core/bloom_filter.hpp

```diff
@@ -0,0 +1,224 @@
+ * to an input, invoke the factory's operator() and use the resulting
+ * HashSource.
+ *
+ * The factory must be kept alive while its HashSource instances remain
+ * in use.
+ */
+class HashSourceFactory
+{
+public:
+  using Bytes     = fetch::byte_array::ConstByteArray;
+  using Function  = std::function<std::vector<std::size_t>(Bytes const &)>;
+  using Functions = std::vector<Function>;
+
+  explicit HashSourceFactory(Functions);
+  ~HashSourceFactory();
+
+  HashSourceFactory()                          = delete;
+  HashSourceFactory(HashSourceFactory const &) = delete;
+  HashSourceFactory(HashSourceFactory &&)      = delete;
+  HashSourceFactory &operator=(HashSourceFactory const &) = delete;
+  HashSourceFactory &operator=(HashSourceFactory &&) = delete;
+
+  HashSource operator()(Bytes const &) const;
+
+private:
+  Functions const hash_functions_;
+};
+
+/*
+ * Represents a sequential application of a HashSourceFactory's hash
+ * functions to a byte array. Outwardly it may be treated as an immutable,
+ * iterable collection of std::size_t.
+ *
+ * Not thread-safe. Not safe to use after the parent HashSourceFactory
+ * had been destroyed.
+ */
+class HashSource
+{
+public:
+  HashSource(HashSource &&) noexcept;
+  HashSource &operator=(HashSource &&) noexcept;
+  ~HashSource();
+
+  HashSource()                   = delete;
+  HashSource(HashSource const &) = delete;
+  HashSource &operator=(HashSource const &) = delete;
+
+  class HashSourceIterator
+  {
+  public:
+    using iterator_category = std::input_iterator_tag;
+    using value_type        = std::size_t const;
+    using difference_type   = std::ptrdiff_t;
+    using pointer           = std::size_t const *;
+    using reference         = std::size_t const &;
+
+    ~HashSourceIterator();
+
+    HashSourceIterator(HashSourceIterator const &);
+    HashSourceIterator(HashSourceIterator &&) noexcept;
+
+    HashSourceIterator &operator=(HashSourceIterator const &);
+    HashSourceIterator &operator=(HashSourceIterator &&) noexcept;
+
+    HashSourceIterator() = delete;
+
+    bool operator!=(HashSourceIterator const &other) const;
+    bool operator==(HashSourceIterator const &other) const;
+
+    HashSourceIterator const operator++(int);
+    HashSourceIterator &     operator++();
+
+    std::size_t operator*() const;
+
+  private:
+    explicit HashSourceIterator(HashSource const *source, std::size_t index);
+
+    std::size_t       hash_index_{};
+    HashSource const *source_{};
+
+    friend class HashSource;
+  };
+
+  HashSourceIterator begin() const;
+  HashSourceIterator end() const;
+  HashSourceIterator cbegin() const;
+  HashSourceIterator cend() const;
+
+private:
+  HashSource(HashSourceFactory::Functions const &, HashSourceFactory::Bytes const &);
+
+  std::size_t getHash(std::size_t index) const;
+
+  std::vector<std::size_t> data_;
+
+  friend class HashSourceFactory;
+};
+
+}  // namespace internal
+
+class BloomFilter
```

Comment:
Style 🚓 : `BloomFilterInterface`

Reply:
Also this should be moved to its own header file to reduce the coupling

Reply:
:+1: